### PR TITLE
Fix SSR issues, new API

### PR DESCRIPTION
**Pull request by andrzejewsky**
- [x] I Accept [Individual Contributor License Agreement](https://github.com/DivanteLtd/next/blob/master/CLA.md)

### CODE REVIEW

File: packages/core/utils/src/event-bus/index.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+const listeners: Listener[] = [];
+let emitted: Event[] = [];
+
+type Handler = (value?: any) => void;
```
Comment (patzick) [inline packages/core/utils/src/event-bus/index.ts]: as this is exposed as external param, should be exported as well
```suggestion
export type Handler = (value?: unknown) => void;
```

File: packages/core/utils/src/event-bus/index.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+interface Listener {
+  eventName: string;
+  handler: Handler;
+}
+
+interface EventBus {
+  emit: (eventName: string, value: any) => void;
+  on: (eventName: string, handler: Handler) => void;
+}
```
Comment (patzick) [inline packages/core/utils/src/event-bus/index.ts]: I believe we don't need this one, methods exports same interface
```suggestion
```

File: packages/core/utils/src/event-bus/index.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+const emit = (eventName: string, value: any): void => {
```
Comment (patzick) [inline packages/core/utils/src/event-bus/index.ts]:
```suggestion
export function emit (eventName: string, value: any): void {
```

File: packages/core/utils/src/event-bus/index.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
+const listeners: Listener[] = [];
+let emitted: Event[] = [];
```
Comment (patzick) [inline packages/core/utils/src/event-bus/index.ts]: we should move it under types declaration, but before exported functions

File: packages/core/utils/src/event-bus/index.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  relatedListeners.forEach((listener) => {
+    listener.handler(value);
+  });
+
+  if (relatedListeners.length === 0) {
+    emitted.push({ eventName, value });
+  }
+};
+
+const on = (eventName: string, handler: Handler): void => {
```
Comment (patzick) [inline packages/core/utils/src/event-bus/index.ts]:
```suggestion
export function on (eventName: string, handler: Handler): void {
```

File: packages/core/utils/src/event-bus/index.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  notConsumed.forEach((event) => {
+    handler(event.value);
+  });
+
+  emitted = emitted.filter(event => event.eventName !== eventName);
+};
+
+const eventBus: EventBus = { emit, on };
+
+export default eventBus;
```
Comment (patzick) [inline packages/core/utils/src/event-bus/index.ts]: we're exporting methods directly
```suggestion
```

File: packages/core/factories/src/useCategoryFactory.ts [hunk 7]
```diff
@@ -10,16 +10,16 @@ export function useCategoryFactory<CATEGORY, CATEGORY_SEARCH_PARAMS>(
   factoryParams: UseCategoryFactoryParams<CATEGORY, CATEGORY_SEARCH_PARAMS>
 ) {
   return function useCategory(cacheId?: string): UseCategory<CATEGORY> {
-    const { state, persistedResource } = usePersistedState(cacheId);
+    const { state } = useSSR(cacheId);
```
Comment (patzick) [inline packages/core/factories/src/useCategoryFactory.ts]: would be nice to have instead
```
const { state, setCache } = useSSR(cacheId) 
```
then it's just call
```
setCache(categories.value)
```
and even cacheId is set at this point internally

File: packages/core/factories/src/useCategoryFactory.ts [hunk 8]
```diff
@@ -10,16 +10,16 @@ export function useCategoryFactory<CATEGORY, CATEGORY_SEARCH_PARAMS>(
...
-      loading.value = true;
-      categories.value = await persistedResource<CATEGORY[]>(
-        factoryParams.categorySearch,
-        params
-      );
+      if (!state) {
+        loading.value = true;
+      }
+      categories.value = await factoryParams.categorySearch(params);
+      eventBus.emit('set-ssr-cache', { key: cacheId, value: categories.value });
```
Comment (patzick) [inline packages/core/factories/src/useCategoryFactory.ts]: after other comment
```suggestion
      setCache(categories.value)
```
wdyt @andrzejewsky @filrak ?

File: packages/core/theme-module/theme/pages/Category.vue [hunk 9]
```diff
@@ -253,26 +255,32 @@ import {
   getCategoryTree,
   getProductVariants
 } from '<%= options.helpers %>';
+import { onSSR } from '@vue-storefront/utils';
 
 export default {
   transition: 'fade',
+  asyncData() {
```
Comment (filrak) [inline packages/core/theme-module/theme/pages/Category.vue]: wtf?

**Review by filrak — CHANGES_REQUESTED** [review]

File: packages/core/theme-module/theme/pages/Category.vue (hunk 9 above)
Comment (andrzejewsky) [reply packages/core/theme-module/theme/pages/Category.vue]: @filrak ić stont, WIP!!

File: packages/core/factories/src/useCategoryFactory.ts [hunk 10]
```diff
@@ -10,16 +10,16 @@ export function useCategoryFactory<CATEGORY, CATEGORY_SEARCH_PARAMS>(
...
 
     const search = async (params: CATEGORY_SEARCH_PARAMS) => {
-      loading.value = true;
-      categories.value = await persistedResource<CATEGORY[]>(
-        factoryParams.categorySearch,
-        params
-      );
+      if (!state) {
+        loading.value = true;
+      }
```
Comment (patzick) [inline packages/core/factories/src/useCategoryFactory.ts]: this is not necessary :) 
```suggestion
      loading.value = true;
```

File: packages/core/factories/src/useProductFactory.ts [hunk 11]
```diff
@@ -17,23 +18,17 @@ export function useProductFactory<PRODUCT, PRODUCT_SEARCH_PARAMS>(
...
-    // const products: Ref<ProductVariant[]> = ref(state || []);\
-    const products: Ref<PRODUCT[]> = ref([]);
-    const totalProducts: Ref<number> = ref(0);
+    const { state, saveCache } = useSSR(cacheId);
+    const products: Ref<PRODUCT[]> = ref(state || []);
+    const totalProducts = ref(0);
     const loading = ref(false);
 
-    const search = async (params: PRODUCT_SEARCH_PARAMS) => {
+    const search = async (params: any) => {
```
Comment (patzick) [inline packages/core/factories/src/useProductFactory.ts]: why is that?
we need to use `PRODUCT_SEARCH_PARAMS` here :)
```suggestion
    const search = async (params: PRODUCT_SEARCH_PARAMS) => {
```

**Review by patzick — CHANGES_REQUESTED** [review]
Tho small changes from me, overriding this looks 👌

File: packages/core/utils/src/event-bus/index.ts (hunk 1 above)
Comment (andrzejewsky) [reply packages/core/utils/src/event-bus/index.ts]: noo this is only for internal usage, it's not used anywhere

File: packages/core/docs/core/utils/server-side-rendering.md [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+## Usage
+
+In the most cases, the composable functions provide SSR support (it is always transparent for developer).
+To build that composable you have to create a new composable-factory with using `useSSR` function.
+
+```js
+import { useSSR } from '@vue-storefront/utils';
+
+export function useExampleFactory(factoryParams) {
```
Comment (filrak) [inline packages/core/docs/core/utils/server-side-rendering.md]: imo factory here is not needed and just adds redundant complexity, just show how to use it with some example use case (using `example` is not the best way to explain stuff as well, show something that can be understood as an idea - simplified product or so)

**Review by filrak — CHANGES_REQUESTED** [review]
Regarding API ( `saveToCache` seems like the right one and if we're saving to cache then we should retrieve `cache`, not state (or the opposite we should save to `state` not cache)

File: packages/core/theme-module/theme/components/RelatedProducts.vue [hunk 13]
```diff
@@ -40,27 +38,18 @@ export default {
...
-  setup({ product }) {
-    const { products, search, loading } = useProduct('related-products');
-    const categories = getProductCategories(product);
-    const relatedProducts = computed(() => getProductVariants(products.value, { masters: true }).filter((prod) => getProductId(prod) !== getProductId(product)));
-
-    if (categories.length > 0) {
-      search({ catId: [categories[0]] });
-    }
-
+  setup() {
```
Comment (filrak) [inline packages/core/theme-module/theme/components/RelatedProducts.vue]: shouldn't setup be always on top?

**Review by filrak — APPROVED** [review]

**Pull request merged**